### DISCUSSION TITLE
generation via build hooks

### DISCUSSION THREAD

**Discussion by chris-kruining** [thesayyn/protoc-gen-ts#115] [Ideas]
as suggested in #113 it might be interesting to introduce build hooks for customizing the output of the generator. I wanted to open this discussion on how we might give this concept shape.

Personally I am leaning into the "simple" solution:
expose a couple of generator classes: interface, client, server, message.

for example: I want to generate another style of client as in my PR
```ts
import { ClassElement } from 'ts';
import { ClientGenerator } from 'proto-gen-ts';

class MyClientGenerator extends ClientGenerator
{
    public createMethod(rootDescriptor, methodDescriptor): ClassElement[]
    {
        return [
            factory.createMethodDeclaration(...),
        ];
    }
}
```

**Comment by thesayyn** [comment]
The idea of having hooks is interesting but I am not sure about how to get this right.  

Want to raise some concerns here;

- How can we have an ABI layer between versions? we tend to change a lot of internals.
- Can a hook alter the AST constructed by protoc-gen-ts,  In case it can, how are we gonna make sure that we use the same typescript version, etc?
- Hooks will have to be able to call some internal APIs such as `type.getTypeReference`, I don't know how to get this right.

This makes me wonder whether it is really worth having this. an alternative could be writing your own plugin which would almost be the same as having a hook.

**Reply by chris-kruining** [reply]
> * How can we have an ABI layer between versions? we tend to change a lot of internals.

build on top of and expose interfaces, that way you have a more stable API

> * Can a hook alter the AST constructed by protoc-gen-ts,  In case it can, how are we gonna make sure that we use the same typescript version, etc?

I wanna say yes, for example, my PR generated a single method for a rpc call, whereas your original code creates 3 to be compatible with the grpc lib

> * Hooks will have to be able to call some internal APIs such as `type.getTypeReference`, I don't know how to get this right.

smart exports I suppose, mostly by determining what is public and what is internal

> This makes me wonder whether it is really worth having this. an alternative could be writing your own plugin which would almost be the same as having a hook.

Same here. The biggest reason for me to still go on with this hooks proposal is because I do not have to maintain a custom plugin, but am still able to customize the output. But looking at the scale of this all, the impact would be HUGE.

**Comment by thesayyn** [comment]
I agree, let's keep this in our pockets while we work on alternatives that would scale well and could be maintained outside of protoc-gen-ts. 

I believe we could have something with `SourceCodeInfo` and `GeneratedCodeInfo`.  these messages map source / generated code span which we could interpret to do partial compilation within separate. i will try to come up with a PoC that does partial generation.
